Regression(269372@main) Crash under SVGPathElement::attributeChanged() after memory pressure

cdumez opened the pull request:
#### ec2d23a0902a7c2e65afbd221a275ce3040b9e1c
<pre>
Regression(269372@main) Crash under SVGPathElement::attributeChanged() after memory pressure
<a href="https://bugs.webkit.org/show_bug.cgi?id=263381">https://bugs.webkit.org/show_bug.cgi?id=263381</a>
rdar://117176058

Reviewed by Said Abou-Hallawa.

We were failing to reset pathSegListCacheSize to 0 after clearing the cache. As a result,
SVGPathElement::attributeChanged() could get in a state where pathSegListCacheSize becomes
greater than maxPathSegListCacheSize even though the cache map is empty. As a result, we
would try to remove `cache.random()` which returns `cache.end()` when the cache is empty.

I also moved the caching logic to a new class for better insulation / clarity.

* LayoutTests/fast/svg/path-element-d-attribute-crash-memory-pressure-expected.txt: Added.
* LayoutTests/fast/svg/path-element-d-attribute-crash-memory-pressure.html: Added.
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::attributeChanged):
(WebCore::SVGPathElement::clearCache):

Canonical link: <a href="https://commits.webkit.org/269547@main">https://commits.webkit.org/269547@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73cea1a69a523c4204e920567dc69e2ef0430b04

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22854 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/812 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23943 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24763 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21153 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23114 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1908 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23381 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23095 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19817 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-api-tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25617 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20698 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26906 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20944 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24755 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/392 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/317 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20501 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5456 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/498 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/SVGPathElement.cpp:
```diff
@@ -76,7 +77,9 @@ void SVGPathElement::attributeChanged(const QualifiedName& name, const AtomStrin
+                        ASSERT(iteratorToRemove != cache.end());
                         ASSERT(pathSegListCacheSize >= iteratorToRemove->value.size());
                         pathSegListCacheSize -= iteratorToRemove->value.size();
                         cache.remove(iteratorToRemove);
```
I think the change in `SVGPathElement::attributeChanged()` is getting bigger and it became harder to read. I think this can be solved by moving the caching to a separate class say `PathSegListCache`. This class should manage adding values to the cache and retrieving existing values from the cache. This will allow early returns which makes the code more readable. The above code may be simplified to be like this:
```
    if (newValue.isEmpty())
        m_pathSegList->baseVal()->updateByteStreamData({ });
    else if (auto* data = PathSegListCache::singleton().get(newValue))
        m_pathSegList->baseVal()->updateByteStreamData(*data);
    else if (m_pathSegList->baseVal()->parse(newValue))
        PathSegListCache::singleton().add(newValue, m_pathSegList->baseVal()->existingPathByteStream().data())
    else
        document().accessSVGExtensions().reportError("Problem parsing d=\"" + newValue + "\"");
```

shallawa commented on Source/WebCore/svg/SVGPathElement.cpp:
```diff
@@ -94,6 +97,7 @@ void SVGPathElement::attributeChanged(const QualifiedName& name, const AtomStrin
 void SVGPathElement::clearCache()
 {
     pathSegListCache().clear();
+    pathSegListCacheSize = 0;
```
The HashMap and the data size can be members of the new class and this function should call 
```
PathSegListCache::singleton().clear();
```

shallawa commented on Source/WebCore/svg/SVGPathElement.cpp:
```diff
@@ -37,15 +37,59 @@ namespace WebCore {
+
+    HashMap<AtomString, SVGPathByteStream::Data> m_cache;
+    uint64_t m_sizeInBytes { 0 };
+    static constexpr uint64_t maxSizeInBytes = 150 * 1024; // 150 Kb.
```
This is the size of data only. The keys can be really long strings. I think the total size of the HashMap can be double this size. Is it expected to have a cache of 300kb for the path data cache?

cdumez replied:
We can have this discussion but this patch is about urgently fixing a crash that needs to be cherry-picked to the branch ASAP. I did not change the cache size in this patch.

The short answer is that the size was chosen was Speedometer 3 content. We can change it but I think this is out of scope for this patch. I already refactored the code as you suggested even though I prefer to keep branch fixes as simple as possible.

shallawa commented on Source/WebCore/svg/SVGPathElement.cpp:
```diff
@@ -37,15 +37,59 @@ namespace WebCore {
+    if (UNLIKELY(newDataSize > (maxSizeInBytes / 2)))
+        return;
+
+    m_sizeInBytes += newDataSize;
```
Should we consider adding small size paths only? Some SVGs like maps use very long path elements. I think this cache will be filled with small number of large paths in this case.

cdumez replied:
Yes, it is worth trying out but needs careful A/B testing on Speedometer 3.

cdumez replied:
I'll A/B test a few tweaks to the cache size and to the limit of the size of items we cache, and follow-up if I can be more conservative without regressing Speedometer 3.

cdumez committed:
Regression(269372@main) Crash under SVGPathElement::attributeChanged() after memory pressure
https://bugs.webkit.org/show_bug.cgi?id=263381
rdar://117176058

Reviewed by Said Abou-Hallawa.

We were failing to reset pathSegListCacheSize to 0 after clearing the cache. As a result,
SVGPathElement::attributeChanged() could get in a state where pathSegListCacheSize becomes
greater than maxPathSegListCacheSize even though the cache map is empty. As a result, we
would try to remove `cache.random()` which returns `cache.end()` when the cache is empty.

I also moved the caching logic to a new class for better insulation / clarity.

* LayoutTests/fast/svg/path-element-d-attribute-crash-memory-pressure-expected.txt: Added.
* LayoutTests/fast/svg/path-element-d-attribute-crash-memory-pressure.html: Added.
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::attributeChanged):
(WebCore::SVGPathElement::clearCache):

Canonical link: https://commits.webkit.org/269547@main
